### PR TITLE
Handle empty lateral average bins

### LINKED ISSUE #547: depth average postprocessor interpolation issue
It looks like if some of your depth average "zones" (bins) don't happen to have any nodes in them, the postprocessor will just return zeros for all values in those bins. That's annoying. I'm opening this issue mostly as a reminder to myself to look in to this later, unless anyone has an answer off the top of their head?

### PR DESCRIPTION
**Pull request by class4kayaker**
Fix #547 by setting the computed average in empty depth bins to nan rather than 0.

Also adds a test case for this behavior. Tester discrepancy is expected first pass on the output, due to machine variation.

### CODE REVIEW

**Comment by class4kayaker** [comment]
Failures noted, will examine non-depth_average tests for reason

**Comment by tjhei** [comment]
Can you let us know when this is ready to review?

**Comment by class4kayaker** [comment]
Waiting on tests, only added material model for now. For the others, I noted that they are using test datafiles as input. If the error is that there are too many lateral bands, would modifying the test input data files or switching to not using lateral average be better. Turning off lateral average is simple, and for some amy produce the same output.

Also, should a failing test case be added for too many bands on Steinberger?

**Comment by tjhei** [comment]
> Also, should a failing test case be added for too many bands on Steinberger for checking future updates?

That would be cool. Do you know how to do that? (see tests/fail.prm)

File: source/material_model/steinberger.cc [hunk 1]
```diff
@@ -435,7 +435,12 @@ namespace aspect
     update()
     {
       if (use_lateral_average_temperature)
-        this->get_lateral_averaging().get_temperature_averages(avg_temp);
+        {
+          this->get_lateral_averaging().get_temperature_averages(avg_temp);
+          for (unsigned int i = 0; i < avg_temp.size(); ++i)
+            Assert(numbers::is_nan(avg_temp[i]),
```
Comment (tjhei) [inline source/material_model/steinberger.cc]: I think you want ``is_finite`` here?

File: source/simulator/lateral_averaging.cc [hunk 2]
```diff
@@ -120,7 +120,19 @@ namespace aspect
     Utilities::MPI::sum(values, this->get_mpi_communicator(), values_all);
 
     for (unsigned int i=0; i<num_slices; ++i)
-      values[i] = values_all[i] / (static_cast<double>(volume_all[i])+1e-20);
+      {
+        if (volume_all[i] > 0.0)
+          {
+            values[i] = values_all[i] / (static_cast<double>(volume_all[i])+1e-20);
```
Comment (tjhei) [inline source/simulator/lateral_averaging.cc]: you no longer need the +1e-20 here, I think.

File: source/simulator/lateral_averaging.cc [hunk 3]
```diff
@@ -120,7 +120,19 @@ namespace aspect
...
+          }
+        else
+          {
+            // Output nan if no quadrature points in depth block
+            this->get_pcout() << "\tWarning: No quadrature points found in depth block " << i
```
Comment (tjhei) [inline source/simulator/lateral_averaging.cc]: @bangerth Do we have a good way to handle warnings? Should a warning only be output once?

File: tests/depth_average_underres.prm [hunk 4]
```diff
@@ -0,0 +1,112 @@
+# Test the 'depth average' postprocessor on a coarse mesh version of
+# convection_box.prm. This test makes no sense since we average over
+# 100 depth intervals but we only compute on a 4x4 mesh -- several of
+# the depth intervals will receive no quadrature points
```
Comment (tjhei) [inline tests/depth_average_underres.prm]: very nice description (and test)!

File: tests/depth_average_underres.prm [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+set Dimension                              = 2
+
+set Use years in output instead of seconds = false
+set End time                               = 0.5
```
Comment (tjhei) [inline tests/depth_average_underres.prm]: from what test is this copied from? (can you add this to the beginning of this file as a comment please)?

Can you set a smaller end time and update the results?

File: source/simulator/lateral_averaging.cc (hunk 3 above)
Comment (bangerth) [inline source/simulator/lateral_averaging.cc]: The question is: "Once we get here, can the code ever continue to produce anything sensible?" If the answer is no, then it should be an `AssertThrow`. If yes, but some of the data is invalid, then we could discuss whether a warning is appropriate, or whether we should also just abort the program. What would a user do with this data?

I'm generally not overly fond of warnings. If we think the input is bad and consequently we can't compute what is requested, then my instinct is to just punt and abort, rather than figure out what subset of the problem we can still solve.

Comment (tjhei) [inline source/simulator/lateral_averaging.cc]: I am fine with removing the warning and let the caller site deal with NaNs in the averaging.

**Comment by class4kayaker** [comment]
Output for all old steinberger tests has been adjusted for not using lateral averaging.

Due to the changes to not use lateral averaging, the steinberger_viscosity test is no longer different from the steinberger_viscosity_adiabatic test. As such, changing back to depth averaging and considering modifying the material model file for less bands may be appropriate. Alternately it may make sense to discard steinberger_viscosity or make it another failing test by changing back to not use lateral averaging.

**Comment by class4kayaker** [comment]
It is still necessary to make the decision on whether to use the current warning approach, or if an empty bin should be considered a fatal failure state. If an empty bin should be considered a failure state, it might be reasonable to scale the iterations for the quadrature by the number of depth bands in that cell to some upper and lower tolerance(possibly subject to cell level). The logic for calculating the number of depth bands might be somewhat more complex, but it would handle the AMR case reasonably and make the fatal failure approach more reasonable.

**Comment by class4kayaker** [comment]
Looking at the output for the material_model_dependencies suggests that the current failure is due to non-convergence for the stokes solver. Would it be reasonable and possible to change to a case that has better convergence behavior given that only the dependency output is apparently of interest?

File: tests/depth_average_underres.prm [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  set Time steps between mesh refinement       = 0
+end
+
+
+subsection Postprocess
+  set List of postprocessors = depth average
+
+  subsection Depth average
+    set Time between graphical output = 0.1
+    set Number of zones = 100
```
Comment (bangerth) [inline tests/depth_average_underres.prm]: reduce this to a smaller number so the test output is smaller

File: tests/material_model_dependencies_steinberger.prm [hunk 7]
```diff
@@ -10,6 +10,7 @@ subsection Material model
     set Latent heat                      = false
     set Compressible                     = true
     set Reference viscosity              = 1e21
+    set Use lateral average temperature for viscosity = false
```
Comment (bangerth) [inline tests/material_model_dependencies_steinberger.prm]: I'd prefer to leave this as is, but reduce the number of depth layers

File: tests/steinberger_viscosity.prm [hunk 8]
```diff
@@ -51,6 +51,7 @@ subsection Material model
     set Latent heat                      = false
     set Compressible                     = true
     set Reference viscosity              = 1e21
+    set Use lateral average temperature for viscosity = false
```
Comment (bangerth) [inline tests/steinberger_viscosity.prm]: Leave it at the default value 'true' if you can. If you can't just reduce the number of depth layers.

File: source/material_model/steinberger.cc [hunk 9]
```diff
@@ -435,7 +435,12 @@ namespace aspect
     update()
     {
       if (use_lateral_average_temperature)
-        this->get_lateral_averaging().get_temperature_averages(avg_temp);
+        {
+          this->get_lateral_averaging().get_temperature_averages(avg_temp);
+          for (unsigned int i = 0; i < avg_temp.size(); ++i)
+            AssertThrow(numbers::is_finite(avg_temp[i]),
+                        ExcMessage("Too many depth bands in lateral average for current refinement"));
```
Comment (bangerth) [inline source/material_model/steinberger.cc]: Maybe change the error message to the following:
```
  In computing depth averages, there is at least one depth band that does not have any quadrature points in it. Consider reducing the number of depth layers for averaging, using the 'BLA BLUB' input parameter.
```

File: source/simulator/lateral_averaging.cc [hunk 10]
```diff
@@ -119,8 +119,25 @@ namespace aspect
...
+          {
+            print_under_res_warning = true;
+            // Output nan if no quadrature points in depth block
+            values[i] = std::numeric_limits<double>::quiet_NaN();
+          }
+      }
+    if (print_under_res_warning)
+      {
+        this->get_pcout() << "\tWarning: No quadrature points found in at least one depth block."
+                          << " Averages are likely underresolved." << std::endl;
```
Comment (bangerth) [inline source/simulator/lateral_averaging.cc]: Maybe reword as follows:
```
 In computing depth averages, there is at least one depth band that does not have any quadrature points in it. 
  Consider reducing the number of depth layers for averaging.
```

**Comment by gassmoeller** [comment]
This PR has been sitting around with working tests for a while. Are we ok with the implemented approach of outputting a warning if bins are empty? @bangerth?
